Mathlib/Data/Real/Sqrt.lean

```diff
@@ -455,4 +455,21 @@ lemma sum_sqrt_mul_sqrt_le (s : Finset ι) (hf : ∀ i, 0 ≤ f i) (hg : ∀ i,
     ∑ i ∈ s, √(f i) * √(g i) ≤ √(∑ i ∈ s, f i) * √(∑ i ∈ s, g i) := by
   simpa [*] using sum_mul_le_sqrt_mul_sqrt _ (fun x ↦ √(f x)) (fun x ↦ √(g x))
 
+/-- **Sedrakyan's lemma**, also known by other names such as Titu's lemma or Engel's form. This is a
+specialization of the Cauchy-Schwarz inequality with the sequences `f n / √(g n)` and `√(g n)`. -/
+theorem sq_sum_div_le_sum_sq_div {s : Finset ι} (f : ι → ℝ) {g : ι → ℝ} (hg : ∀ i ∈ s, 0 < g i) :
+    (∑ n ∈ s, f n) ^ 2 / ∑ n ∈ s, g n ≤ ∑ n ∈ s, (f n) ^ 2 / g n := by
```

Comment:
It seems a little unfortunate that this lemma only works for `Real` and not `NNReal`. Presumably it's also true for `Rat`?

Reply:
It should be true for any linearly ordered field, the same way Cauchy-Schwarz is. But if CS isn't in Mathlib then I don't think there's much case in trying to generalize this inequality.

Reply:
(I don't actually know how you'd Sedrakyan for linear orders fields, other than to work in the quadratic closure. For Cauchy-Schwarz the proof with bilinear forms generalizes directly.)

Reply:
> But if CS isn't in Mathlib then I don't think there's much case in trying to generalize this inequality.

I don't really understand what you mean by that. Cauchy-Schwarz is in mathlib in the general case, but your application of the lemma requires a square root function which obviously doesn't work for Rat. So the loss of generality seems to be a function of your proof, not of the lemmas in mathlib.

If the generalization is awkward a TODO would suffice.

Reply:
> requires a square root function which obviously doesn't work for Rat

What about taking a real closure (do we have this in mathlib?), or simply taking a real quadratic extension?

Reply:
Ah, sorry. I think I saw the earlier theorems in this file (the ones about square roots) were specific to real numbers and assumed the same for `Finset.sum_mul_sq_le_sq_mul_sq`.

I've shot my shot at [Stack Exchange](https://math.stackexchange.com/q/5001456/718671) asking for a proof of Sedrakyan without linear ordered fields. If anyone can come up with that then we'll be able to avoid a lot of field shenanigans :)

Reply:
I think what's actually needed is a more general version of `sum_mul_sq_le_sq_mul_sq`, where `f i ^ 2` and `g i ^ 2` are replaced by arbitrary nonnegative values (not necessarily squares, to avoid needing square roots); given the more general Cauchy-Schwarz in semirings, this result should follow. That is, something like

```lean
lemma sum_sq_le_sum_mul_sum_of_sq_eq_mul (s : Finset ι) {t f g : ι → R}
    (ht : ∀ i, t i ^ 2 = f i * g i) (hf : ∀ i, 0 ≤ f i) (hg : ∀ i, 0 ≤ g i) :
    (∑ i ∈ s, t i) ^ 2 ≤ (∑ i ∈ s, f i) * ∑ i ∈ s, g i := by
  sorry
```

(for `[LinearOrderedCommSemiring R] [ExistsAddOfLE R]`, like the original). The existing proof of `sum_mul_sq_le_sq_mul_sq` uses `two_mul_le_add_sq` as the essential use of "squares are nonnegative"; instead you probably need a lemma of the form `2 * t ≤ a + b` given `t ^ 2 = a * b`, `0 ≤ a` and `0 ≤ b` (don't know if we already have it, but it follows in an ordered semiring by squaring both sides).

Reply:
I've pushed a proof of this auxiliary lemma. I'll try using it to generalize Cauchy-Schwarz and then Sedrakyan.

Reply:
Proofs successfully generalized!